mlir/lib/Target/LLVMIR/DebugImporter.cpp

```diff
@@ -346,3 +385,20 @@ StringAttr DebugImporter::getStringAttrOrNull(llvm::MDString *stringNode) {
     return StringAttr();
   return StringAttr::get(context, stringNode->getString());
 }
+
+DistinctAttr DebugImporter::getOrCreateDistinctID(llvm::DINode *node) {
+  DistinctAttr &id = nodeToDistinctAttr[node];
+  if (!id)
+    id = DistinctAttr::create(UnitAttr::get(context));
+  return id;
+}
+
+function_ref<DIRecursiveTypeAttrInterface(DistinctAttr)>
+DebugImporter::getRecSelfConstructor(llvm::DINode *node) {
+  using CtorType = function_ref<DIRecursiveTypeAttrInterface(DistinctAttr)>;
+  return TypeSwitch<llvm::DINode *, CtorType>(node)
+      .Case<llvm::DICompositeType>([&](auto *concreteNode) {
```

Comment:
```suggestion
      .Case([&](llvm::DICompositeType *concreteNode) {
```

Wouldn't this make it a bit simpler? Maybe MSVC is somehow falling apart due to this. This might additionally also help in simplifying the inner template magic even further.

Reply:
Yeah you're right. I was trying to see if I could get it to be generic, so we don't need duplicate cases just for each type that supports recursion, but maybe it's not worth it if each one is just one line 😂.